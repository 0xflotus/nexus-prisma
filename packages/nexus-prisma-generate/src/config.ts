import * as fs from 'fs'
import * as yaml from 'js-yaml'
import * as path from 'path'
import generateCRUDSchemaString from 'prisma-generate-schema'

export function findDatamodelAndComputeSchema(): string {
  const configPath = findPrismaConfigFile()

  if (!configPath) {
    throw new Error('Could not find `prisma.yml` file')
  }

  let definition: null | any = null

  try {
    const file = fs.readFileSync(configPath, 'utf-8')
    definition = yaml.safeLoad(file)
  } catch (e) {
    throw new Error(`Yaml parsing error in ${configPath}: ${e.message}`)
  }

  if (!definition.datamodel) {
    throw new Error('Missing `datamodel` property in prisma.yml file')
  }

  const typeDefs = getTypesString(
    definition.datamodel,
    path.dirname(configPath),
  )

  return generateCRUDSchemaString(typeDefs)
}

function findPrismaConfigFile(): string | null {
  let definitionPath: string | null = path.join(process.cwd(), 'prisma.yml')

  if (fs.existsSync(definitionPath)) {
    return definitionPath
  }

  definitionPath = path.join(process.cwd(), 'prisma', 'prisma.yml')

  if (fs.existsSync(definitionPath)) {
    return definitionPath
  }

  return null
}

function getTypesString(datamodel: string | string[], definitionDir: string) {
  const typesPaths = datamodel
    ? Array.isArray(datamodel)
      ? datamodel
      : [datamodel]
    : []

  let allTypes = ''

  typesPaths.forEach(unresolvedTypesPath => {
    const typesPath = path.join(definitionDir, unresolvedTypesPath!)
    if (fs.existsSync(typesPath)) {
      const types = fs.readFileSync(typesPath, 'utf-8')
      allTypes += types + '\n'
    } else {
      throw new Error(
        `The types definition file "${typesPath}" could not be found.`,
      )
    }
  })

  return allTypes
}

export function findRootDirectory(): string {
  const cwd = process.cwd()
  const tsConfig = findConfigFile(cwd, 'tsconfig.json')

  if (tsConfig) {
    return path.dirname(tsConfig)
  }

  const packageJson = findConfigFile(cwd, 'package.json')

  if (packageJson) {
    return path.dirname(packageJson)
  }

  return cwd
}

function findConfigFile(
  searchPath: string,
  configName = 'package.json',
): string | undefined {
  return forEachAncestorDirectory(searchPath, ancestor => {
    const fileName = path.join(ancestor, configName)
    return fs.existsSync(fileName) ? fileName : undefined
  })
}

/** Calls `callback` on `directory` and every ancestor directory it has, returning the first defined result. */
function forEachAncestorDirectory<T>(
  directory: string,
  callback: (directory: string) => T | undefined,
): T | undefined {
  while (true) {
    const result = callback(directory)
    if (result !== undefined) {
      return result
    }

    const parentPath = path.dirname(directory)
    if (parentPath === directory) {
      return undefined
    }

    directory = parentPath
  }
}

export function getImportPathRelativeToOutput(
  importPath: string,
  outputDir: string,
): string {
  let relativePath = path.relative(path.dirname(outputDir), importPath)

  if (!relativePath.startsWith('.')) {
    relativePath = './' + relativePath
  }

  // remove .ts or .js file extension
  relativePath = relativePath.replace(/\.(ts|js)$/, '')

  // remove /index
  relativePath = relativePath.replace(/\/index$/, '')

  // replace \ with /
  relativePath = relativePath.replace(/\\/g, '/')

  return relativePath
}
