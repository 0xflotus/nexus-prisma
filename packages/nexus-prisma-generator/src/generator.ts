import * as fs from 'fs'
import { GraphQLSchema } from 'graphql'
import { join } from 'path'
import { ISDL } from 'prisma-datamodel'
import { generateCRUDSchemaFromInternalISDL } from 'prisma-generate-schema'
import { renderTypes, renderDatamodelInfo } from './types-generator'
import {
  findDatamodelAndComputeSchema,
  readPrismaYml,
} from './types-generator/config'
import { promisify } from 'util'

const copyFileAsync = promisify(fs.copyFile)

const INPUT_LIBRARY_DIR = join(__dirname, 'library')
const DIRS = ['blocks', 'definitions', 'graphqlToNexus', 'generated']
const STATIC_FILES = [
  'camelcase',
  'graphql',
  'index',
  'throw',
  'typesHelpers',
  'validate',
  'builder',
  'resolver',
  'types',
  'utils',
  join('blocks', 'objectType'),
  join('blocks', 'inputObjectType'),
  join('blocks', 'extendType'),
  join('definitions', 'index'),
  join('definitions', 'objectType'),
  join('definitions', 'inputObjectType'),
  join('definitions', 'enumType'),
  join('definitions', 'extendType'),
  join('graphqlToNexus', 'index'),
  join('graphqlToNexus', 'common'),
  join('graphqlToNexus', 'enumType'),
  join('graphqlToNexus', 'inputObjectType'),
  join('graphqlToNexus', 'objectType'),
]
const GENERATED_TYPES = 'generated/types.d.ts'
const DATAMODEL_INFO = 'generated/datamodelInfo.js'

export async function generate(
  prismaYmlPath: string,
  opts?: { clientDir?: string; outputPath?: string },
) {
  if (!opts) {
    opts = {}
  }
  if (!opts.clientDir) {
    opts.clientDir = '@generated/prisma-client'
  }
  if (!opts.outputPath) {
    opts.outputPath = join(__dirname, '..', '..', '@generated/nexus-prisma')
  }

  const { datamodel, schema } = readDatamodel(prismaYmlPath)
  const outputDirs = DIRS.map(dirPath => join(opts!.outputPath!, dirPath))
  ;[...outputDirs, opts.outputPath].forEach(path => {
    try {
      // Create the output directories if needed (mkdir -p)
      fs.mkdirSync(path, { recursive: true })
    } catch (e) {
      if (e.code !== 'EEXIST') throw e
    }
  })

  await copyStaticFiles(opts.outputPath)

  renderDynamicFiles(
    datamodel,
    schema,
    opts.clientDir || '@generated/prisma-client',
    opts.outputPath,
  )
}

function readDatamodel(prismaYmlPath: string) {
  const prisma = readPrismaYml(prismaYmlPath)

  const { datamodel, databaseType } = findDatamodelAndComputeSchema(
    prisma.configPath,
    prisma.config,
  )
  const schema = generateCRUDSchemaFromInternalISDL(datamodel, databaseType)

  return {
    datamodel,
    schema,
  }
}

async function copyStaticFiles(output: string) {
  const jsFilesPromises = STATIC_FILES.map(filePath =>
    copyFileAsync(
      join(INPUT_LIBRARY_DIR, filePath + '.js'),
      join(output, filePath + '.js'),
    ),
  )
  const dTSPromises = STATIC_FILES.map(filePath =>
    copyFileAsync(
      join(INPUT_LIBRARY_DIR, filePath + '.d.ts'),
      join(output, filePath + '.d.ts'),
    ),
  )

  return Promise.all([...jsFilesPromises, ...dTSPromises])
}

function renderDynamicFiles(
  datamodel: ISDL,
  schema: GraphQLSchema,
  clientDir: string,
  output: string,
) {
  const datamodelInfoPath = join(output, DATAMODEL_INFO)

  const datamodelInfo = renderDatamodelInfo(
    datamodel,
    schema,
    'module.exports =',
  )

  fs.writeFileSync(datamodelInfoPath, datamodelInfo)

  renderTypes({
    output: join(output, GENERATED_TYPES),
    prismaClientDir: clientDir,
    schema,
  })
}
