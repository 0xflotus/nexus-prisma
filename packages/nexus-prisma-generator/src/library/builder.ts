import { buildClientSchema, GraphQLNamedType } from 'graphql'
import { core } from 'nexus'
import { graphqlTypeToNexus } from './graphqlToNexus'
import { InternalDatamodelInfo, InternalPrismaSchemaConfig } from './types'

export class PrismaSchemaBuilder extends core.SchemaBuilder {
  private datamodelInfo: InternalDatamodelInfo

  constructor(protected config: InternalPrismaSchemaConfig) {
    super(config)

    const requiredDatamodelInfo = require('./generated/datamodelInfo')

    this.datamodelInfo = {
      ...requiredDatamodelInfo,
      schema: buildClientSchema(requiredDatamodelInfo.schema),
    }
  }

  protected missingType(typeName: string): GraphQLNamedType {
    const datamodelInfo = this.getDatamodelInfo()
    const type = datamodelInfo.schema.getType(typeName)

    if (type) {
      return graphqlTypeToNexus(this, type, this.config.photon, datamodelInfo)
    }

    return super.missingType(typeName)
  }

  public getConfig() {
    return this.config
  }

  public getDatamodelInfo() {
    return this.datamodelInfo
  }
}

export function isPrismaSchemaBuilder(obj: any): obj is PrismaSchemaBuilder {
  return obj && obj instanceof PrismaSchemaBuilder
}
