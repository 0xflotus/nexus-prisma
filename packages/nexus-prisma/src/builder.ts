import { buildClientSchema, GraphQLNamedType, GraphQLSchema } from 'graphql'
import { core } from 'nexus'
import { graphqlTypeToNexus } from './graphqlToNexus'
import { PrismaSchemaConfig } from './types'

export class PrismaSchemaBuilder extends core.SchemaBuilder {
  private nexusPrismaSchema: {
    uniqueFieldsByModel: Record<string, string[]>
    schema: GraphQLSchema
  }

  constructor(protected config: PrismaSchemaConfig) {
    super(config)

    this.nexusPrismaSchema = {
      uniqueFieldsByModel: this.config.prisma.datamodelInfo.uniqueFieldsByModel,
      schema: buildClientSchema(this.config.prisma.datamodelInfo.schema),
    }
  }

  protected missingType(typeName: string): GraphQLNamedType {
    const type = this.getNexusPrismaSchema().schema.getType(typeName)

    if (type) {
      return graphqlTypeToNexus(
        this,
        type,
        this.config.prisma.client,
        this.config.prisma.datamodelInfo.uniqueFieldsByModel,
      )
    }

    return super.missingType(typeName)
  }

  public getConfig() {
    return this.config
  }

  public getNexusPrismaSchema() {
    return this.nexusPrismaSchema
  }
}

export function isPrismaSchemaBuilder(obj: any): obj is PrismaSchemaBuilder {
  return obj && obj instanceof PrismaSchemaBuilder
}
