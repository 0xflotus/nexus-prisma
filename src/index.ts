import { makeSchemaWithMetadata } from 'nexus'
import { PrismaSchemaConfig } from './types'
import { SchemaBuilder, Metadata } from 'nexus/dist/core'

export { prismaObjectType } from './prisma'

class PrismaSchemaBuilder extends SchemaBuilder {
  constructor(metadata: Metadata, config: PrismaSchemaConfig) {
    super(metadata, config)
  }

  public getConfig() {
    return this.config
  }
}

export function buildPrismaSchema(options: PrismaSchemaConfig) {
  const { schema, metadata } = makeSchemaWithMetadata(
    options,
    PrismaSchemaBuilder,
  )

  // Only in development envs do we want to worry about regenerating the
  // schema definition and/or generated types.
  const {
    shouldGenerateArtifacts = process.env.NODE_ENV !== 'production',
  } = options

  if (shouldGenerateArtifacts) {
    // Generating in the next tick allows us to use the schema
    // in the optional thunk for the typegen config
    metadata.generateArtifacts(schema)
  }

  return schema
}
