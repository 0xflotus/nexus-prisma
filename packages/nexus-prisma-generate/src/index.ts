#!/usr/bin/env node

import { writeFileSync, existsSync, mkdirSync } from 'fs'
import { EOL } from 'os'
import { join, relative, dirname } from 'path'
import * as meow from 'meow'
import {
  extractTypes,
  GraphQLTypeField,
  GraphQLTypeObject,
  GraphQLTypes,
  GraphQLTypeArgument,
  GraphQLType,
  GraphQLEnumObject,
} from './source-helper'
import { findDatamodelAndComputeSchema, findRootDirectory } from './config'

const cli = meow(
  `
    nexus-prisma-generate prisma-client-dir output

    > Generate the TypeScript types for nexus-prisma

    -----
    
    Inputs should be relative to the root of your project

    \`prisma-client-dir\`: Path to your prisma-client directory (eg: ./generated/nexus-prisma/)
    \`output\`: Path to where you want to output the typings (eg: ./generated/nexus-prisma.ts)
`,
)

main(cli)

function main(cli: meow.Result) {
  let [prismaClientDir, output] = cli.input

  if (!prismaClientDir || !existsSync(prismaClientDir)) {
    console.log('No valid `prisma-client-dir` was found')
    process.exit(1)
  }

  if (!output || !output.endsWith('.ts')) {
    console.log('No valid `output` was found. Must point to a .ts file')
    process.exit(1)
  }

  const rootPath = findRootDirectory()
  const schema = findDatamodelAndComputeSchema()
  const types = extractTypes(schema)
  const typesToRender = render(
    types,
    getImportPathRelativeToOutput(prismaClientDir, output),
  )

  // Create the output directories if needed (mkdir -p)
  mkdirSync(dirname(output), { recursive: true })

  const outputPath = join(rootPath, output)

  writeFileSync(outputPath, typesToRender)
  console.log(`Types generated at ${output}`)
}

function render(types: GraphQLTypes, prismaClientPath: string) {
  const objectTypes = types.types.filter(t => t.type.isObject)
  const inputTypes = types.types.filter(t => t.type.isInput)
  const enumTypes = types.enums

  return `\
// GENERATED TYPES FOR PRISMA PLUGIN. /!\\ DO NOT EDIT MANUALLY

import {
  ArgDefinition,
  ContextValue,
  RootValue,
} from 'nexus/dist/types'
import { GraphQLResolveInfo } from 'graphql'

import * as prisma from '${prismaClientPath}'

${objectTypes.map(renderType).join(EOL)}

${inputTypes.map(renderInputType).join(EOL)}

${renderEnumTypes(enumTypes)}

export interface PluginTypes {
  fields: {
${objectTypes
  .map(type => `    ${type.name}: ${getExposableObjectsTypeName(type)}`)
  .join(EOL)}
  }
  fieldsDetails: {
${objectTypes
  .map(type => `    ${type.name}: ${getTypeObjectName(type)}`)
  .join(EOL)}
  }
  enumTypesNames: ${getEnumTypesName()}
}

declare global {
  interface GraphQLNexusGen extends PluginTypes {}
}
  `
}

function renderType(type: GraphQLTypeObject) {
  return `\
// Types for ${type.name}

${renderFields(type)}

${renderFieldsArgs(type)}

${renderTypeFieldDetails(type)}
`
}

function renderTypeFieldDetails(type: GraphQLTypeObject) {
  return `\
export interface ${getTypeObjectName(type)}<GenTypes = GraphQLNexusGen> {
${type.fields
  .map(
    field => `\
  ${field.name}: {
    args: ${
      field.arguments.length > 0
        ? `Record<${getTypeFieldArgName(type, field)}, ArgDefinition>`
        : '{}'
    }
    description: string
    list: ${field.type.isArray || false}
    nullable: ${!field.type.isRequired}
    resolve: (
      root: RootValue<GenTypes, "${type.name}">,
      args: ${renderResolverArgs(field)},
      context: ContextValue<GenTypes>,
      info?: GraphQLResolveInfo
    ) => ${renderResolverReturnType(field)};
  }`,
  )
  .join(EOL)}
}
  `
}

function renderResolverArgs(field: GraphQLTypeField) {
  return `{\
 ${field.arguments
   .map(
     arg => `${arg.name}${arg.type.isRequired ? '' : '?'}: ${getTSType(arg)}`,
   )
   .join(', ')} }\
  `
}

function renderFields(type: GraphQLTypeObject) {
  return `\
type ${getExposableObjectsTypeName(type)} =
  | ${getExposableFieldsTypeName(type)}
${type.fields
  .map(
    f =>
      `  | { name: '${f.name}', args?: ${
        f.arguments.length > 0
          ? `${getTypeFieldArgName(type, f)}[] | false`
          : '[] | false'
      }, alias?: string  } `,
  )
  .join(EOL)}

type ${getExposableFieldsTypeName(type)} =
${type.fields.map(f => `  | '${f.name}'`).join(EOL)}
`
}

function renderFieldsArgs(type: GraphQLTypeObject) {
  return `\
${type.fields
  .filter(field => field.arguments.length > 0)
  .map(field => renderFieldArg(type, field))
  .join(EOL)}
  `
}

function renderFieldArg(type: GraphQLTypeObject, f: GraphQLTypeField) {
  return `\
type ${getTypeFieldArgName(type, f)} =
${f.arguments.map(arg => `  | '${arg.name}'`).join(EOL)}`
}

function getTSType(graphqlType: GraphQLTypeField | GraphQLTypeArgument) {
  const graphqlToTypescript: Record<string, string> = {
    String: 'string',
    Boolean: 'boolean',
    ID: 'string',
    Int: 'number',
    Float: 'number',
    DateTime: 'string',
  }

  let returnType = ''

  if (graphqlType.type.isScalar) {
    returnType = graphqlToTypescript[graphqlType.type.name]
  } else if (graphqlType.type.isInput) {
    returnType = getInputTypeName(graphqlType.type)
  } else {
    returnType = `prisma.${graphqlType.type.name}`
  }

  if (graphqlType.type.isArray) {
    returnType += '[]'
  }

  if (!graphqlType.type.isRequired) {
    returnType += ' | null'
  }

  return returnType
}

function renderResolverReturnType(field: GraphQLTypeField) {
  const returnType = getTSType(field)

  return `Promise<${returnType}> | ${returnType}`
}

function renderInputType(input: GraphQLTypeObject) {
  return `\
export interface ${getInputTypeName(input)} {
${input.fields
  .map(
    field =>
      `  ${field.name}${field.type.isRequired ? '' : '?'}: ${getTSType(field)}`,
  )
  .join(EOL)}
}
  `
}

function renderEnumTypes(enums: GraphQLEnumObject[]): string {
  return `\
export type ${getEnumTypesName()} =
${enums.map(e => `  | '${e.name}'`).join(EOL)}
  `
}

function getExposableFieldsTypeName(type: GraphQLTypeObject) {
  return `${type.name}Fields`
}

function upperFirst(s: string) {
  return s.replace(/^\w/, c => c.toUpperCase())
}

function getTypeFieldArgName(type: GraphQLTypeObject, field: GraphQLTypeField) {
  return `${type.name}${upperFirst(field.name)}Args`
}

function getExposableObjectsTypeName(type: GraphQLTypeObject) {
  return `${type.name}Object`
}

function getTypeObjectName(type: GraphQLTypeObject) {
  return `${type.name}FieldDetails`
}

function getInputTypeName(type: GraphQLTypeObject | GraphQLType) {
  return `${type.name}`
}

function getEnumTypesName(): string {
  return 'enumTypesNames'
}

export function getImportPathRelativeToOutput(
  importPath: string,
  outputDir: string,
): string {
  let relativePath = relative(dirname(outputDir), importPath)

  if (!relativePath.startsWith('.')) {
    relativePath = './' + relativePath
  }

  // remove .ts or .js file extension
  relativePath = relativePath.replace(/\.(ts|js)$/, '')

  // remove /index
  relativePath = relativePath.replace(/\/index$/, '')

  // replace \ with /
  relativePath = relativePath.replace(/\\/g, '/')

  return relativePath
}
