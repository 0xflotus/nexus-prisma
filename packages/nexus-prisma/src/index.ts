import { GraphQLSchema } from 'graphql'
import { core } from 'nexus'
import { PrismaSchemaBuilder } from './builder'
import { PrismaSchemaConfig } from './types'

export * from './definitions'

interface PrismaSchemaConfigRequiredTypes extends PrismaSchemaConfig {
  types: any
}

function validateOptions(options: PrismaSchemaConfig): void {
  if (!options.prisma) {
    throw new Error(
      'Missing `prisma` property in `makePrismaSchema({ prisma: { ... } })`',
    )
  }

  if (!options.prisma.datamodelInfo) {
    throw new Error(
      'Missing `prisma.datamodelInfo` property in `makePrismaSchema({ prisma: { datamodelInfo: ... } })`',
    )
  }

  if (
    !options.prisma.datamodelInfo.uniqueFieldsByModel ||
    !options.prisma.datamodelInfo.clientPath ||
    !options.prisma.datamodelInfo.schema
  ) {
    throw new Error(
      'Invalid `prisma.datamodelInfo` property. This should be imported from the `nexus-prisma-generate` output directory',
    )
  }

  if (!options.prisma.client) {
    throw new Error(
      'Missing `prisma.client` property in `makePrismaSchema({ prisma: { client: ... } })`',
    )
  }

  if (
    typeof options.prisma.client !== 'function' &&
    (!options.prisma.client.$exists || !options.prisma.client.$graphql)
  ) {
    throw new Error(
      `\
Invalid \`prisma.client\` property in \`makePrismaSchema({ prisma: { client: ... } })\`.
This should either be an instance of the generated prisma-client, or a function that returns the prisma-client instance from your GraphQL server context
`,
    )
  }
}

export function makePrismaSchema(options: PrismaSchemaConfig): GraphQLSchema {
  validateOptions(options)

  const builder = new PrismaSchemaBuilder(options)

  if (!options.types) {
    options.types = []
  }

  const { schema } = core.makeSchemaInternal(
    options as PrismaSchemaConfigRequiredTypes,
    builder,
  )

  // Only in development envs do we want to worry about regenerating the
  // schema definition and/or generated types.
  const {
    shouldGenerateArtifacts = Boolean(
      !process.env.NODE_ENV || process.env.NODE_ENV === 'development',
    ),
  } = options

  if (shouldGenerateArtifacts) {
    // Generating in the next tick allows us to use the schema
    // in the optional thunk for the typegen config
    new core.TypegenMetadata(options).generateArtifacts(schema).catch(e => {
      console.error(e)
    })
  }

  return schema
}
