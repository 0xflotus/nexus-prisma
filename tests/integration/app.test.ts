import * as cp from 'child_process'
import * as path from 'path'
import * as nexusBuilder from 'nexus/dist/builder'
import * as nexusPrisma from '../../src'
import * as fs from 'fs-extra'

beforeAll(async () => {
  await fs.mkdirp(relative('generated/nexus-types'))
})

afterAll(async () => {
  // NOTE
  // Comment out this line if you want to play around with the integration
  // test app in VSCode with actual IDE TS type check feedback.
  await fs.remove(relative('generated'))
})

// IDEA Future tests?
// - show we gracefully handle case of photon import failing

it('integrates together', async () => {
  // Run Prisma generation:
  // - Photon JS Client
  cp.execSync('../../node_modules/.bin/prisma2 generate', {
    cwd: __dirname,
  })

  // Run Nexus generation:
  // - Core TypeScript Declaration
  // - Prisma TypeScript Declaration
  // - GraphQL Schema
  //
  // NOTE
  // This also acts as a bootstrap phase, producing types that
  // enable subsequent type checks. A user currently has to figure
  // this part out on their own more or less.
  //
  await nexusPrisma.generateTypes({
    typegenPath: generatedPath(`nexus-types/prisma.d.ts`),
  })

  await nexusBuilder.generateSchema({
    types: [
      require('./app'),
      nexusPrisma.nexusPrismaPlugin({ photon: ctx => ctx.photon }),
    ],
    outputs: {
      typegen: generatedPath(`nexus-types/core.d.ts`),
      schema: generatedPath(`schema.graphql`),
    },
  })

  // Assert the app type checks. In effect this is testing that our
  // typegen works. Snapshot our generated files for manual correctness
  // tracking.
  //
  expect(relative('tsconfig.json')).toTypeCheck()
  expect(await getGenerated('schema.graphql')).toMatchSnapshot()
  expect(await getGenerated('nexus-types/prisma.d.ts')).toMatchSnapshot()
  expect(await getGenerated('nexus-types/core.d.ts')).toMatchSnapshot()
})

async function getGenerated(relPath: string): Promise<string> {
  return fs.readFile(path.join(generatedPath(relPath))).then(b => b.toString())
}

async function getRelative(relPath: string): Promise<string> {
  return fs.readFile(path.join(relative(relPath))).then(b => b.toString())
}

function generatedPath(relPath: string): string {
  return path.join(relative('generated'), relPath)
}

function relative(relPath: string): string {
  return path.join(__dirname, relPath)
}
