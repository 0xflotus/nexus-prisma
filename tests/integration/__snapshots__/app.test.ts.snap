// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integrates together 1`] = `
"### This file was autogenerated by GraphQL Nexus
### Do not make changes to this file directly


type A {
  hash: String!
  id: Int!
}

type Query {
  ok: Boolean!
}
"
`;

exports[`integrates together 2`] = `
"import * as photon from '@generated/photon';
import { core } from 'nexus';
// Types helpers
  type ModelNameExistsInGraphQLType<
  ReturnType extends any
> = ReturnType extends core.GetGen<'objectNames'> ? true : false;

type NexusPrismaScalarOpts = {
  alias?: string;
};

type Pagination = {
  first?: boolean;
  last?: boolean;
  before?: boolean;
  after?: boolean;
  skip?: boolean;
};

type RootObjectTypes = Pick<
  core.GetGen<'rootTypes'>,
  core.GetGen<'objectNames'>
>;

type IsSubset<A, B> = keyof A extends never
  ? false
  : B extends A
  ? true
  : false;

type OmitByValue<T, ValueType> = Pick<
  T,
  { [Key in keyof T]: T[Key] extends ValueType ? never : Key }[keyof T]
>;

type GetSubsetTypes<ModelName extends any> = keyof OmitByValue<
  {
    [P in keyof RootObjectTypes]: ModelName extends keyof ModelTypes
      ? IsSubset<RootObjectTypes[P], ModelTypes[ModelName]> extends true
        ? RootObjectTypes[P]
        : never
      : never;
  },
  never
>;

type SubsetTypes<ModelName extends any> = GetSubsetTypes<
  ModelName
> extends never
  ? \`ERROR: No subset types are available. Please make sure that one of your GraphQL type is a subset of your t.model('<ModelName>')\`
  : GetSubsetTypes<ModelName>;

type DynamicRequiredType<ReturnType extends any> = ModelNameExistsInGraphQLType<
  ReturnType
> extends true
  ? { type?: SubsetTypes<ReturnType> }
  : { type: SubsetTypes<ReturnType> };

type GetNexusPrismaInput<
  ModelName extends any,
  MethodName extends any,
  InputName extends 'filtering' | 'ordering'
> = ModelName extends keyof NexusPrismaInputs
  ? MethodName extends keyof NexusPrismaInputs[ModelName]
    ? NexusPrismaInputs[ModelName][MethodName][InputName]
    : never
  : never;

type NexusPrismaRelationOpts<
  ModelName extends any,
  MethodName extends any,
  ReturnType extends any
> = GetNexusPrismaInput<
  // If GetNexusPrismaInput returns never, it means there are no filtering/ordering args for it. So just use \`alias\` and \`type\`
  ModelName,
  MethodName,
  'filtering'
> extends never
  ? {
      alias?: string;
    } & DynamicRequiredType<ReturnType>
  : {
      alias?: string;
      filtering?:
        | boolean
        | Partial<
            Record<
              GetNexusPrismaInput<ModelName, MethodName, 'filtering'>,
              boolean
            >
          >;
      ordering?:
        | boolean
        | Partial<
            Record<
              GetNexusPrismaInput<ModelName, MethodName, 'ordering'>,
              boolean
            >
          >;
      pagination?: boolean | Pagination;
    } & DynamicRequiredType<ReturnType>;

type IsScalar<TypeName extends any> = TypeName extends core.GetGen<
  'scalarNames'
>
  ? true
  : false;

type NexusPrismaFields<ModelName extends keyof NexusPrismaTypes> = {
  [MethodName in keyof NexusPrismaTypes[ModelName]]: NexusPrismaMethod<
    ModelName,
    MethodName,
    IsScalar<NexusPrismaTypes[ModelName][MethodName]> // Is the return type a scalar?
  >;
};

type NexusPrismaMethod<
  ModelName extends keyof NexusPrismaTypes,
  MethodName extends keyof NexusPrismaTypes[ModelName],
  IsScalar extends boolean,
  ReturnType extends any = NexusPrismaTypes[ModelName][MethodName]
> = IsScalar extends true // If scalar
  ? (opts?: NexusPrismaScalarOpts) => NexusPrismaFields<ModelName> // Return optional scalar opts
  : ModelNameExistsInGraphQLType<ReturnType> extends true // If model name has a mapped graphql types
  ? (
      opts?: NexusPrismaRelationOpts<ModelName, MethodName, ReturnType>
    ) => NexusPrismaFields<ModelName> // Then make opts optional
  : (
      opts: NexusPrismaRelationOpts<ModelName, MethodName, ReturnType>
    ) => NexusPrismaFields<ModelName>; // Else force use input the related graphql type -> { type: '...' }

type GetNexusPrismaMethod<
  TypeName extends string
> = TypeName extends keyof NexusPrismaMethods
  ? NexusPrismaMethods[TypeName]
  : <CustomTypeName extends keyof ModelTypes>(
      typeName: CustomTypeName
    ) => NexusPrismaMethods[CustomTypeName];

type GetNexusPrisma<
  TypeName extends string,
  ModelOrCrud extends 'model' | 'crud'
> = ModelOrCrud extends 'model'
  ? TypeName extends 'Mutation'
    ? never
    : TypeName extends 'Query'
    ? never
    : GetNexusPrismaMethod<TypeName>
  : ModelOrCrud extends 'crud'
  ? TypeName extends 'Mutation'
    ? GetNexusPrismaMethod<TypeName>
    : TypeName extends 'Query'
    ? GetNexusPrismaMethod<TypeName>
    : never
  : never;
  

// Generated
interface ModelTypes {
  A: photon.A
}
  
interface NexusPrismaInputs {
  Query: {
    as: {
  filtering: 'id' | 'AND' | 'OR' | 'NOT'
  ordering: 'id'
}

  },
    A: {


  }
}

interface NexusPrismaTypes {
  Query: {
    as: 'A'
    a: 'A'

  },
  Mutation: {
    createOneA: 'A'
    deleteOneA: 'A'
    updateOneA: 'A'
    updateManyA: 'BatchPayload'
    deleteManyA: 'BatchPayload'

  },
  A: {
    id: 'Int'

}
}

interface NexusPrismaMethods {
  A: NexusPrismaFields<'A'>
  Query: NexusPrismaFields<'Query'>
  Mutation: NexusPrismaFields<'Mutation'>
}
  

declare global {
  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;

exports[`integrates together 3`] = `
"/**
 * This file was automatically generated by GraphQL Nexus
 * Do not make changes to this file directly
 */





declare global {
  interface NexusGenCustomOutputProperties<TypeName extends string> {
    crud: NexusPrisma<TypeName, 'crud'>
    model: NexusPrisma<TypeName, 'model'>
  }
}

declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
}

export interface NexusGenRootTypes {
  A: { // root type
    hash: string; // String!
    id: number; // Int!
  }
  Query: {};
  String: string;
  Int: number;
  Float: number;
  Boolean: boolean;
  ID: string;
}

export interface NexusGenAllTypes extends NexusGenRootTypes {
}

export interface NexusGenFieldTypes {
  A: { // field return type
    hash: string; // String!
    id: number; // Int!
  }
  Query: { // field return type
    ok: boolean; // Boolean!
  }
}

export interface NexusGenArgTypes {
}

export interface NexusGenAbstractResolveReturnTypes {
}

export interface NexusGenInheritedFields {}

export type NexusGenObjectNames = \\"A\\" | \\"Query\\";

export type NexusGenInputNames = never;

export type NexusGenEnumNames = never;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = \\"Boolean\\" | \\"Float\\" | \\"ID\\" | \\"Int\\" | \\"String\\";

export type NexusGenUnionNames = never;

export interface NexusGenTypes {
  context: any;
  inputTypes: NexusGenInputs;
  rootTypes: NexusGenRootTypes;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  allTypes: NexusGenAllTypes;
  inheritedFields: NexusGenInheritedFields;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractResolveReturn: NexusGenAbstractResolveReturnTypes;
}"
`;
